### PR TITLE
Implement Action Runner on Manager

### PR DESCRIPTION
**Pull request by jgrund**
This patch implements the IML ActionRunner as a manager side plugin. The
plugin will bookkeep active sessions and any in-flight RPCs to IML
agents.

Other IML services will integrate with this service to make and receive
calls via HTTP over unix domain socket.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

### CODE REVIEW

File: iml-services/src/services/action_runner.rs [hunk 1]
```diff
@@ -3,32 +3,188 @@
...
+            } => match sessions.lock().get(&fqdn) {
+                Some(held_session) if held_session == &session_id => {
+                    log::info!("good session {:?}/{:?}", fqdn, session_id);
+
+                    match rpcs.lock().get_mut(&session_id) {
+                        Some(rs) => {
+                            let result: Result<ActionResult, String> =
+                                serde_json::from_value(body).unwrap();
+
+                            let result = result.unwrap();
```
Comment (johnsonw) [inline iml-services/src/services/action_runner.rs]:
```suggestion
let {id, result} = result.unwrap();
```

File: iml-services/src/services/action_runner.rs [hunk 2]
```diff
@@ -3,32 +3,188 @@
...
-            .map_err(|e| {
-                log::error!("An error occured: {:?}", e);
+                    })
             })
+            .map_err(exit.trigger_fn())
+            .map_err(|e| log::error!("An error occured: {:?}", e));
+
+        tokio::spawn(consume_from_agent);
+
+        // let consume_from_manager = iml_rabbit::connect_to_rabbit().and_then(move |client| {});
```
Comment (johnsonw) [inline iml-services/src/services/action_runner.rs]: Still needed?

File: iml-services/src/services/action_runner/main.rs [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    // fn abort_session(fqdn: Fqdn) {}
+
+    // fn replay_rpcs<'a>(client: TcpClient, rpcs: &'a mut Rpcs<'a>, old_id: &Id, new_id: &'a Id) {
+    //     if let Some(xs) = rpcs.remove(old_id) {
+    //         for action_in_flight in xs.values() {
+    //             // tokio::spawn(send_agent_message(client, msg));
+    //         }
+    //         rpcs.insert(new_id, xs);
+    //     };
+    // }
```
Comment (utopiabound) [inline iml-services/src/services/action_runner/main.rs]: Dead code?

File: iml-services/src/services/action_runner/receiver.rs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                    fqdn
+                ));
+
+                action_in_flight.complete(msg).unwrap();
+            }
+        }
+    }
+}
+
+pub fn hande_agent_data(
```
Comment (johnsonw) [inline iml-services/src/services/action_runner/receiver.rs]:
```suggestion
pub fn handle_agent_data
```

File: iml-services/src/services/action_runner/main.rs (hunk 3 above)
Comment (jgrund) [reply iml-services/src/services/action_runner/main.rs]: Removed in latest push.

File: iml-services/src/services/action_runner.rs (hunk 2 above)
Comment (jgrund) [reply iml-services/src/services/action_runner.rs]: Removed in latest push.

File: iml-services/src/services/action_runner.rs (hunk 1 above)
Comment (jgrund) [reply iml-services/src/services/action_runner.rs]: Removed in latest push.

File: iml-services/src/services/action_runner.rs [hunk 5]
```diff
@@ -3,32 +3,188 @@
...
+
 fn main() {
     env_logger::init();
 
+    let (exit, valve) = tokio_runtime_shutdown::shared_shutdown();
+
+    let sessions: Shared<Sessions> = Arc::new(Mutex::new(HashMap::new()));
+    let rpcs: Shared<Rpcs> = Arc::new(Mutex::new(HashMap::new()));
+
+    // fn abort_session(fqdn: Fqdn) {}
```
[the remark this answers is not in the document]
Comment (jgrund) [reply iml-services/src/services/action_runner.rs]: Removed in latest push.

**Pull request merged**